### PR TITLE
Add __all__ for a few distributed modules plus a little typing

### PR DESCRIPTION
**Pull request by kumpera**
This handles distributed_c10d, which is massive and ddp_comm_hooks.

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
Seems that this whole PR is now contained in https://github.com/pytorch/pytorch/pull/84139#pullrequestreview-1094481963, is that correct?

**Comment by kumpera** [comment]
> Seems that this whole PR is now contained in [#84139 (review)](https://github.com/pytorch/pytorch/pull/84139#pullrequestreview-1094481963), is that correct?

yes, but this is smaller change that I'll rebase the other one on top once it's merged

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -44,25 +61,56 @@
...
+        Work
+    ]
+    for type in _public_types_to_change_module:
+        type.__module__ = "torch.distributed.distributed_c10d"
+_export_c_types()
+
 try:
     from torch._C._distributed_c10d import ProcessGroupMPI
+    ProcessGroupMPI.__module__ = "torch.distributed.distributed_c10d"
+    __all__ += ["ProcessGroupMPI"]
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:91]: Ideally users should not be doing things like `c10d.ProcessGroupNCCL` to directly create PGs, but unfortunately it is already an exposed API.

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**